CLUSTER.md

```diff
@@ -0,0 +1,514 @@
+This approach assumes that the rate of change for the global configuration is (at most) on the same order of magnitude as 
+the rate of change for the cluster topology (adding / removing nodes & slot ownership). Settings which change very frequently 
+cannot be handled by the centralized control plane as it would quickly be overwhelmed.
+By propagating the individual node 'global config' epoch as part of the status messages to the TD we can expose when a certain configuration
+change has been received by all nodes.
+Node local config changes (to the relevant global settings) would be overridden every time a change is propagated through
+the cluster.
+
+Another possibility would be to leverage node level distribution mechanism like unsharded Pub/Sub.
+- Would require some kind of reliability guarantee that isn't present in the current/proposed implementation.
+- Has the benefit of much higher scalability, supporting Lua-like Functions usage.
+
+### Reference Implementation
+
+Our reference implementation uses a single Redis module that contains both the Failover Coordinator & Topology Director functionality
+and incorporates an open-source Raft library (The module is written in Rust with the Raft lib in C).
+The interface between the various components (Topology Director, Failover Coordinator, Nodes) uses Protobuf serialized over RESP.
+Other than the very minimal data node logic, all Failover Coordinator/Topology Director logic is external to the main Redis event loop.
+
+It is important to stress that it is perfectly possible to implement the Flotilla specification using something other 
+than Redis+modules. 
+
+#### Road to Production
+
+Once this specification goes through a review process by the community, we intend to push a PoC implementation of the spec
+to the main Redis repo (with the same BSD license as Redis) and enhance it until it reaches GA quality. When this is achieved, 
+we will use the next major version to deprecate the existing cluster implementation and release the two side-by-side (with a compilation
+flag to control which one is used).
+
+As a big TODO - we need to create a troubleshooting "Everything is broken, what do!" section once we have a more concrete
+implementation. Said implementation will also need to provide "force override" flags for various operations which will allow an admin
+to override some Flotilla validations in order to restore availability ASAP.
+Specifically, need to address:
+- How to fix a situation where there's an FC level issue which prevents failovers from taking place.
+- Handling rejected topology changes due to nodes which haven't acknowledged a previous change by overriding the validation 
+  or removing the unresponsive nodes. 
+
+<!--
+Some future topics:
+
+* Define an official RESP-based wire interface
+* Define a function call interface for a "pluggable cluster.c"
+* Describe how ClusterV2 will be implemented based on that.
+-->
+
+#### Potential Enhancements
```

Comment:
I vaguely thought we discussed this at somepoint, but what is the authentication between data nodes and TD/FC?